Model/src/main/java/org/apidb/apicommon/model/report/ai/expression/DailyCostMonitor.java

```diff
@@ -0,0 +1,141 @@
+    _costPerInputToken = getNumberProp(wdkModel, PLAN_COST_PER_1M_INPUT_TOKENS_PROP_NAME) / 1000000;
+    _costPerOutputToken = getNumberProp(wdkModel, PLAN_COST_PER_1M_OUTPUT_TOKENS_PROP_NAME) / 1000000;
+  }
+
+  private double getNumberProp(WdkModel wdkModel, String propName) throws WdkModelException {
+    try {
+      return Double.parseDouble(Optional
+          .ofNullable(wdkModel.getProperties().get(propName))
+          .orElseThrow(() -> new WdkModelException("WDK property '" + propName + "' has not been set.")));
+    }
+    catch (NumberFormatException e) {
+      throw new WdkModelException("WDK property '" + propName + "' must be a number.");
+    }
+  }
+
+  public boolean isCostExceeded() {
+    return updateAndGetCost(EMPTY_COST) > _maxDailyDollarCost;
+  }
+
+  public void updateCost(Optional<CompletionUsage> usage) {
+    updateAndGetCost(usage.orElse(EMPTY_COST));
+  }
+
+  public double updateAndGetCost(CompletionUsage usageCost) {
+    try (DirectoryLock lock = new DirectoryLock(_costMonitoringDir, DEFAULT_TIMEOUT_MILLIS, DEFAULT_POLL_FREQUENCY_MILLIS)) {
+
+      // read current values from file
+      JSONObject previousJson = readAccumulatedCostFile();
+      String previousDate = previousJson.getString(JSON_DATE_PROP);
+      double previousCost = previousJson.getDouble(JSON_COST_PROP);
+
+      // calculate cost of the current usage
+      double additionalCost =
+          (usageCost.promptTokens() * _costPerInputToken) +
+          (usageCost.completionTokens() * _costPerOutputToken);
+
+      // reset cost to zero if date has rolled over to the next day
+      String newDate = getCurrentDateString();
+      if (!newDate.equals(previousDate)) {
+        // reset cost
+        previousCost = 0;
+      }
+
+      double newCost = previousCost + additionalCost;
+
+      // only write file if necessary
+      if (!newDate.equals(previousDate) || newCost != previousCost) {
+
+        try (Writer out = new FileWriter(_costMonitoringFile.toFile())) {
+          out.write(new JSONObject()
+              .put(JSON_DATE_PROP, newDate)
+              .put(JSON_COST_PROP, newCost)
+              .toString()
+          );
+        }
+      }
+
+      return newCost;
+    }
+    catch (Exception e) {
+      throw new WdkRuntimeException("Unable to update AI expression cost", e);
+    }
+  }
+
+  private static String getCurrentDateString() {
+    return FormatUtil.STANDARD_DATE_FORMAT.format(LocalDate.now());
+  }
+
+  private JSONObject readAccumulatedCostFile() {
+    if (!_costMonitoringFile.toFile().exists()) {
+      return new JSONObject()
+          .put(JSON_DATE_PROP, getCurrentDateString())
+          .put(JSON_COST_PROP, 0);
+    }
+    try (Reader in = new FileReader(_costMonitoringFile.toFile())) {
+      return new JSONObject(IoUtil.readAllChars(in));
+    }
+    catch (IOException e) {
+      throw new WdkRuntimeException("Unable to read AI expression cost file: " + _costMonitoringFile);
```

Comment:
I'm assuming this will catch a file reading error or a JSON syntax error?

Do we want to throw or do we want to try to rescue the situation?

The only scenarios I can imagine are
1. one of us edits the file (duh!)
2. the disk is full and the file is malformed (but then we have much bigger problems)

Reply:
Added JSONException to the catch